SEP-2549: TTL for List Results (#2549)
modelcontextprotocol/modelcontextprotocol: docs/specification/draft/server/utilities/caching.mdx

---

> title: Caching

Reviewer: We should probably add a security section to this saying that public cacheScope can lead to primitives being exposed from one user to another and that servers should (a) make sure that cache scope correctly reflects visibility scope and (b) recommend that server authors have appropriate per-primitive access controls and don't rely entirely on caching -- i.e. don't assume that just because admins tools are only shown to admins that you don't also need to check permissions on every tool call (why: because tools lists are not ACLs in the first place, and caching can lead to staleness).

Author: added

---

<div id="enable-section-numbers" />

The Model Context Protocol (MCP) supports caching on list and
read results. This allows clients to cache responses and reduce unnecessary re-fetching.
Caching is complementary to [change notifications](#interaction-with-notifications)&mdash;both
mechanisms can coexist.

## Cacheable Results

Servers MUST include caching hints on results returned by the following operations:

- `tools/list`
- `prompts/list`
- `resources/list`
- `resources/templates/list`
- `resources/read`

## Cacheable Model

Cacheable Results in MCP use two fields to provide caching hints to clients:

- The <b>Time-to-live (TTL) Field</b>,`ttlMs`, indicates how long the client MAY consider the result fresh.
- The <b>Cache Scope Field</b>,`cacheScope`, indicates the intended scope of the cached response, either `"public"` or `"private"`.

> ### Time-to-Live (TTL) Field

Reviewer 2: Temporal is in TC39 and perfectly support Ms:

https://tc39.es/proposal-temporal/docs/duration.html

<Note>
  TTL is a **freshness hint**, not a guarantee. Servers MAY change the
  underlying data before the TTL expires. The TTL tells the client how long it
  can reasonably avoid re-fetching, not how long the data is guaranteed to
  remain unchanged.
</Note>

#### Freshness Calculation

A client records the local time at which the response was received (`t_received`). The
response is considered **fresh** while:

```
now < t_received + ttlMs
```

Once the TTL expires, the response is **stale** and the client **SHOULD** re-fetch on
next access.

Clients **SHOULD NOT** treat TTL as a polling interval that triggers automatic background
refetches. The TTL is a freshness hint: the client checks freshness when it needs the
data, and re-fetches only if stale. Implementations that do choose to poll **MUST**
apply jitter and backoff.

Clients **MAY** re-fetch before the TTL expires if they have reason to believe the data
has changed (e.g., receiving an unexpected error on a tool call indicating the method was
not found or the parameters were invalid).

Clients **MAY** serve stale responses if errors occur during re-fetching (e.g., network
issues, server downtime).

> ### Cache Scope Field

Reviewer: What about per-capability caching? We started discussing in Transports but never figured out the right solution, i.e. suppose I have a `tools/list` with public cacheScope -- if a client asks for `tools/list` with elicitation capability / or tasks extension, can I share that with another use that doesn't have those capabilities? I expect some servers will filter `tools/list` based on capabilities.

We may say that these list should be private, but I think that just moves the problem: should Alice share a `tools/list` cache across Claude Code and Claude Desktop where only one e.g. has support for MCP Apps?

I think we probably want to include capability hash in the cache key for both cache scopes?

Author: This is a good problem for us to solve, I don't know that we have time to solve for this release, as discussed in Core Maintainers.

Will punt to future releases, and perhaps provide guidance in developer docs.

The `cacheScope` field controls who may cache a response, analogous to HTTP
`Cache-Control: public` vs `Cache-Control: private`.

| Value       | Meaning                                                                                                                                                 |
| ----------- | ------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `"public"`  | The response does not contain user-specific data. Any client, shared gateway, or caching proxy **MAY** store and serve the cached response to any user. |

> | `"private"` | The response contains user-specific data. Only the requesting user's client **MAY** cache it. Shared caches **MUST NOT** serve it to a different user.  |

Reviewer: Is "different user" a well-defined concept in MCP? Concretely, can I share the cache across the same OAuth principal with different scopes? What if Alice fetching tool with admin-scoped token, and then starts a new conversation with read-only token? She'll still get the admin tools. Feels like this could be dangerous. cc @pcarleton for details. Not sure what's the right thing to do here. Cache per auth token?

(I really want this btw! Just want to make sure we get it right).

Reviewer: Need to also consider things like client credentials, or stdio-with-api-key-in-an-env forms of identity.

Reviewer 3: I agree that cache per auth would be the right barrier here. Don't try to tell if it's the same user, just need to refetch if you re-auth (in whatever form).

Author: Great will update spec language and thanks for the check.  The thing I'm trying to solve for is some MCP servers tools/list operations are behind auth even when they are the same set of tools for everyone. 

From a security perspective should we allow these Servers to return public? or should we just say if its behind auth it needs to always be private.

Reviewer: No, I agree we need auth'd servers to be able to return `public`. Just wanted to make sure that when we return `private` we have the right semantics.

Reviewer 3: How about this?

```suggestion
| `"private"` | The response contains private data that is not meant to be shared between callers. Cached responses **MAY** be reused for the same authorization context. Caches **MUST NOT** be shared across authorization contexts (e.g. a different access token requires a different cache).  |
```

#### Choosing a Cache Scope

- **`"public"`** is appropriate for lists of tools, prompts, and resource templates that

>   are identical for all users.

Reviewer: Wording change here. Current wording could read that tools list etc. are always identical for all users.
```suggestion
- **`"public"`** is appropriate for lists of tools, prompts, and resource templates when
  they are identical for all users.
```

---
title: Caching
---

<div id="enable-section-numbers" />

The Model Context Protocol (MCP) supports caching on list and

> read results. This allows clients to cache responses and reduce unnecessary re-fetching.

Reviewer: ```suggestion
resource read results. This allows clients to cache responses and reduce unnecessary re-fetching.
```

    Client->>Server: tools/list
    Server-->>Client: { tools: [...], ttlMs: 300000 }
```

## Interaction with Pagination

When a list result is [paginated](/specification/draft/server/utilities/pagination), each
page is an independently cacheable response&mdash;consistent with how HTTP
`Cache-Control` treats paginated resources.

- Each page response carries its own `ttlMs` value. The freshness clock for each page
  starts at the time that page was received.
- Servers **MAY** return different `ttlMs` values on different pages (e.g., a longer TTL
  for early pages of a stable list, a shorter TTL for the final page).
- When a cached page expires, the client **SHOULD** re-fetch that page using its cursor.
- There is no cross-page consistency guarantee. If the underlying data changes between
  page fetches, clients may observe duplicates or gaps.
- Clients that require a consistent snapshot of the full list **SHOULD** re-fetch from
  the beginning (without a cursor).
- If a cursor becomes invalid (e.g., the server returns an error for a previously valid
  cursor), the client **SHOULD** discard all cached pages and re-fetch from the
  beginning.

## Security Considerations

> A `cacheScope` of `"public"` indicates that the response does not contain user-specific data and can be safely shared. However, servers MUST ensure that responses marked as `"public"` can lead to primitives being exposed across users and auth context.

Reviewer: Struggling to parse the second sentence. Is it meant to say "servers MUST **be aware** that responses ...."

Author: good flag, updated and included an example to try to make this clearer.

Server implementors should ensure that:

- the `cacheScope` correctly reflects the intended visibility of the primitive.
- Server authors MUST apply appropriate per-primitive access controls, and MUST NOT rely on

>   `cacheScope` alone to prevent unauthorized access to primitives.

Reviewer: The "Server implementors should ensure that" and then the second bullet "Server authors MUST" doesn't really follow.

```suggestion
Server implementors:

- should ensure that the `cacheScope` correctly reflects the intended visibility of the primitive.
- MUST apply appropriate per-primitive access controls, and MUST NOT rely on
  `cacheScope` alone to prevent unauthorized access to primitives.
```
